### PR TITLE
Add service events for azure cloud provider

### LINKED ISSUE #63155: Azure DNS label mismatch leads to endless failed attempts to ensure load balancer
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Ensuring Load Balancer times out and repeats endlessly

**What you expected to happen**:

And error of some kind that will prevent helm from waiting until timeout

**How to reproduce it (as minimally and precisely as possible)**:

Create an azure cluster with `--dns-label-prefix foo`

Create a kubernetes deployment with a service with the annotation:
`service.beta.kubernetes.io/azure-dns-label-name: bar`
Relevant PR:
https://github.com/kubernetes/kubernetes/pull/47849

**Anything else we need to know?**:

So basically this issue occurs when there is a mismatch between the dns label prefix upon cluster creation and the value used with the annotation

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9.6
- Cloud provider or hardware configuration: Azure Container Services (AKS)
- Install tools: helm

/sig azure
/sig network

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

azure-cloud-provider events are now added for easily identify the underground errors of Azure API.

Before this PR, only `timed out waiting for the condition` on the service events. Users couldn't know what's wrong unless login to master and check kube-controller-manager logs:

```
$ kubectl describe svc nginx
...
Events:
  Type     Reason                         Age               From                  Message
  ----     ------                         ----              ----                  -------
  Normal   EnsuringLoadBalancer           5s (x3 over 28s)  service-controller    Ensuring load balancer
  Warning  CreatingLoadBalancerFailed     0s (x3 over 24s)  service-controller    Error creating load balancer (will retry): failed to ensure load balancer for service default/nginx: timed out waiting for the condition
```

After this PR, the details errors would also show in service's events:

```
$ kubectl describe svc nginx
...
Events:
  Type     Reason                         Age               From                  Message
  ----     ------                         ----              ----                  -------
  Normal   EnsuringLoadBalancer           5s (x3 over 28s)  service-controller    Ensuring load balancer
  Warning  CreateOrUpdatePublicIPAddress  0s (x6 over 26s)  azure-cloud-provider  network.PublicIPAddressesClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="DnsRecordInUse" Message="DNS record bar.southeastasia.cloudapp.azure.com is already used by another public IP." Details=[]
  Warning  CreatingLoadBalancerFailed     0s (x3 over 24s)  service-controller    Error creating load balancer (will retry): failed to ensure load balancer for service default/nginx: timed out waiting for the condition
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63155

**Special notes for your reviewer**:

**Release note**:

```release-note
Service events are now added in azure-cloud-provider for easily identify the underground errors of Azure API.

Action required: The following clusterrole and clusterrolebinding should be applied:

    kind: List
    apiVersion: v1
    items:
    - apiVersion: rbac.authorization.k8s.io/v1
      kind: ClusterRole
      metadata:
        labels:
          kubernetes.io/cluster-service: "true"
        name: system:azure-cloud-provider
      rules:
      - apiGroups: [""]
        resources: ["events"]
        verbs:
        - create
        - patch
        - update
    - apiVersion: rbac.authorization.k8s.io/v1
      kind: ClusterRoleBinding
      metadata:
        labels:
          kubernetes.io/cluster-service: "true"
        name: system:azure-cloud-provider
      roleRef:
        apiGroup: rbac.authorization.k8s.io
        kind: ClusterRole
        name: system:azure-cloud-provider
      subjects:
      - kind: ServiceAccount
        name: azure-cloud-provider
        namespace: kube-system

If the clusterrole with same has already been provisioned (e.g. for accessing azurefile secrets), then the above yaml should be merged togather, e.g.

    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRole
    metadata:
      labels:
        kubernetes.io/cluster-service: "true"
      name: system:azure-cloud-provider
    rules:
    - apiGroups: [""]
      resources: ["events"]
      verbs:
      - create
      - patch
      - update
    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRoleBinding
    metadata:
      labels:
        kubernetes.io/cluster-service: "true"
      name: system:azure-cloud-provider
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: system:azure-cloud-provider
    subjects:
    - kind: ServiceAccount
      name: azure-cloud-provider
      namespace: kube-system
    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRole
    metadata:
      name: system:azure-persistent-volume-binder 
      labels:
        kubernetes.io/cluster-service: "true"
    rules:
    - apiGroups: ['']
      resources: ['secrets']
      verbs:     ['get','create']
    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRoleBinding
    metadata:
      name: system:azure-persistent-volume-binder
      labels:
        kubernetes.io/cluster-service: "true"
    roleRef:
      kind: ClusterRole
      apiGroup: rbac.authorization.k8s.io
      name: system:azure-persistent-volume-binder 
    subjects:
    - kind: ServiceAccount
      name: persistent-volume-binder
      namespace: kube-system
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
@feiskyer It's conflict now

**Comment by feiskyer** [comment]
@andyzhangx Solved now. PTAL

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Comment by feiskyer** [comment]
/assign @liggitt

@liggitt Could you help to take a look at the PR?

File: pkg/cloudprovider/providers/azure/azure.go [hunk 1]
```diff
@@ -383,7 +391,12 @@ func parseConfig(configReader io.Reader) (*Config, error) {
 }
 
 // Initialize passes a Kubernetes clientBuilder interface to the cloud provider
-func (az *Cloud) Initialize(clientBuilder controller.ControllerClientBuilder) {}
+func (az *Cloud) Initialize(clientBuilder controller.ControllerClientBuilder) {
+	az.kubeClient = clientBuilder.ClientOrDie("azure-cloud-provider")
```
Comment (liggitt) [inline pkg/cloudprovider/providers/azure/azure.go]: cc @tallclair @deads2k 
for more cloud provider clients

I know we don't want roles for these getting added to bootstrap policy. what client should they use to speak to the API?

**Comment by liggitt** [comment]
cloud-specific roles should not be part of the bootstrap policy installed on all clusters (xref https://github.com/kubernetes/kubernetes/issues/66628 for clearing out the one that slipped in accidentally)

I'm actually out of the office, redirecting to @tallclair / @mikedanese / @deads2k for guidance on what client to use for reporting events instead (or alternately how to set up the role to give the cloudprovider-specific client event permissions)

**Comment by feiskyer** [comment]
@liggitt Thanks for the comments. 

> I'm actually out of the office, redirecting to @tallclair / @mikedanese / @deads2k for guidance on what client to use for reporting events instead (or alternately how to set up the role to give the cloudprovider-specific client event permissions)

@tallclair @deads2k @mikedanese Could you please take a look at this?

**Comment by feiskyer** [comment]
kindly ping @tallclair @deads2k @mikedanese

File: pkg/cloudprovider/providers/azure/azure.go (hunk 1 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure.go]: @liggitt Didn't get the response yet from @tallclair @deads2k. Any suggestion of candidate role should be used for such cases?

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 2]
```diff
@@ -489,6 +489,13 @@ func ClusterRoles() []rbacv1.ClusterRole {
 				rbacv1helpers.NewRule("create").Groups(certificatesGroup).Resources("certificatesigningrequests/selfnodeclient").RuleOrDie(),
 			},
 		},
+		{
+			// a role making azure cloud provider to send events
+			ObjectMeta: metav1.ObjectMeta{Name: "system:azure-cloud-provider"},
```
Comment (deads2k) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: This is the role that @liggitt noted shouldn't go into core.

**Comment by deads2k** [comment]
While there is still discussion over exactly which org cloud providers should be in, they all seem to agree that k/k isn't the right spot.  Given that view, I don't think it makes sense to use a service account from the `kube-system` namespace.  I'd suggest choosing a namespace where you'd eventually want the azure cloud provider to live and create a client from there.  For the clusterrole and clusterrolebinding, I'd suggest providing a manifest that can be applied.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go (hunk 2 above)
Comment (feiskyer) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: @deads2k Thanks. Then let me remove this and add a release note on it. This is a user-fencing change.

**Comment by andyzhangx** [comment]
@brendandburns @khenidak do you have any opinion about this PR? We are going to make this PR into v1.12, it's very useful for debugging, while clusterrole and rolebinding needs to be added into acs-engine and AKS otherwise there will be error event in service description.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**